server/test/sc/server/plugins/TestGame.kt

```diff
@@ -0,0 +1,103 @@
+import org.slf4j.LoggerFactory
+import sc.api.plugins.IGameState
+import sc.api.plugins.exceptions.TooManyPlayersException
+import sc.framework.plugins.ActionTimeout
+import sc.framework.plugins.Player
+import sc.framework.plugins.RoundBasedGameInstance
+import sc.protocol.responses.ProtocolMessage
+import sc.server.helpers.TestTeam
+import sc.server.helpers.WinReason
+import sc.shared.PlayerScore
+import sc.shared.ScoreCause
+import sc.shared.WelcomeMessage
+import sc.shared.WinCondition
+import java.util.*
+
+class TestGame : RoundBasedGameInstance<TestPlayer?>(TestPlugin.TEST_PLUGIN_UUID) {
+    private val state = TestGameState()
+    override fun onRoundBasedAction(fromPlayer: Player, data: ProtocolMessage) {
+        /*
+         * NOTE: Checking if right player sent move was already done by
+         * {@link sc.framework.plugins.RoundBasedGameInstance#onAction(Player, Object)}.
+         * There is no need to do it here again.
+         */
+
+        if (data is TestMove) {
+            data.perform(state)
+            next(if (state.currentPlayer === TestTeam.RED) state.red else state.blue)
+        }
+    }
+
+    override fun checkWinCondition(): WinCondition? {
+        return if (this.round > 1) {
+            logger.info("Someone won")
+            WinCondition(if (state.state % 2 == 0) TestTeam.RED else TestTeam.BLUE, WinReason.ROUND_LIMIT_FREE_FIELDS)
+        } else null
+    }
+
+    @Throws(TooManyPlayersException::class)
+    override fun onPlayerJoined(): Player {
+        if (players.size < 2) {
+            return if (players.isEmpty()) {
+                state.red = TestPlayer(TestTeam.RED)
+                players.add(state.red)
+                state.red
+            } else {
+                state.blue = TestPlayer(TestTeam.BLUE)
+                players.add(state.blue)
+                state.blue
+            }
+        }
+        throw TooManyPlayersException()
+    }
+
+    val testPlayers: List<TestPlayer?>
+        get() = players
+
+    override fun getPlayerScores(): List<PlayerScore> = emptyList()
+
+    override fun getCurrentState(): IGameState = state
+
+    override fun onPlayerLeft(player: Player, cause: ScoreCause?) {
+        // this.players.remove(player);
+        logger.debug("Player left $player")
+        val result = generateScoreMap()
+        result[player] = PlayerScore(cause, "Spieler hat das Spiel verlassen.", 0)
+        notifyOnGameOver(result)
+    }
+
+    override fun onPlayerLeft(player: Player) {
+        onPlayerLeft(player, ScoreCause.LEFT)
+    }
+
+    override fun getScoreFor(p: TestPlayer?): PlayerScore? {
+        return PlayerScore(true, "Spieler hat gewonnen.")
+    }
+
+    override fun loadFromFile(file: String) {}
+    override fun loadFromFile(file: String, turn: Int) {}
+    override fun loadGameInfo(gameInfo: Any) {}
+    override fun getWinners(): List<Player> = emptyList()
+
+    override fun getPlayers(): List<Player> {
+        return ArrayList<Player>(players)
+    }
```

Comment:
please change this

Reply:
23f38480

Reply:
not quite what I meant ;)
we shouldn't use `ArrayList<Player>` at all here, but simply `listOf` so Kotlin can decide which list is appropriate

Reply:
ah, but that doesn't work. `players` is a list already, so `listOf(players) returns a list of lists.
`.flatten()` doesn't infer the type properly, thus `ArrayList<Player>(players)` was the most reasonable change without using something like `.map()`

Reply:
`toList` should work, but whatever ^^

Reply:
```Type mismatch: inferred type is List<TestPlayer?> but List<Player> was expected```

Reply:
ah, but the problem seemed to be the nullability part.
`testPlayers.toList()` works